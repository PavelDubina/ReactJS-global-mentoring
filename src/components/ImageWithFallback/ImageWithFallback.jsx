import React, { useState } from 'react'
import PropTypes from 'prop-types'
import Poster from '../../static/images/fallback-image.jpg'

export const ImageWithFallback = ({ src = '' }) => {
  const [error, setError] = useState(false)
  const onImageError = () => setError(true)
  const imgSrc = !error ? src : Poster
  return <img onError={onImageError} src={imgSrc} width="9" height="16" alt="poster" />
}

ImageWithFallback.propTypes = {
  src: PropTypes.string,
}
