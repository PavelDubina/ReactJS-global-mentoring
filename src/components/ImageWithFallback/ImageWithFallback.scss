img {
  position: absolute;
  top: 0;
  min-width: 100%;
  height: auto;
  object-fit: cover;
}
