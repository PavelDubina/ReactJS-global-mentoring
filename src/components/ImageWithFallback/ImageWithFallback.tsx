import React, { useState, useEffect } from 'react'
import Poster from '../../../public/fallback-image.jpg'

type ImageWithFallbackProps = {
  src?: string
  id?: string | string[]
  onClick?: () => void
}

export const ImageWithFallback: React.FC<ImageWithFallbackProps> = React.memo(({ src = Poster, onClick, id }) => {
  const [error, setError] = useState(false)
  const onImageError = () => setError(true)
  const imgSrc = !error ? src : Poster
  useEffect(() => {
    if (id) setError(false)
  }, [src, id])
  return <img onError={onImageError} onClick={onClick} src={imgSrc} width="9" height="16" alt="poster" />
})
