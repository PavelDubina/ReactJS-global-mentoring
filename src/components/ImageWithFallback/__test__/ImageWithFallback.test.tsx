import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { ImageWithFallback } from '../ImageWithFallback'

describe('ImageWithFallback', () => {
  type Prop = { onClick?: () => void; src?: any }
  const getComponent = ({ onClick = jest.fn(), src }: Prop) => render(<ImageWithFallback onClick={onClick} src={src} />)
  it('Should renders correctly with empty src', () => {
    expect(getComponent({}).asFragment()).toMatchSnapshot()
  })
  it('Should called onClick', () => {
    const onClick = jest.fn()
    const { getByAltText } = getComponent({ onClick })
    const image = getByAltText('poster')
    fireEvent.click(image)
    expect(onClick).toBeCalled()
  })
  it('After onError event src changes to mock src', () => {
    const { getByAltText } = getComponent({ src: '' })
    const image: any = getByAltText('poster')
    fireEvent.error(image)
    expect(image.src).toBe('http://localhost/test-file-stub')
  })
})
