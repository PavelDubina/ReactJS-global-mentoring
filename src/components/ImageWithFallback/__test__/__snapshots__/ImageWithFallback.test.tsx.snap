// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageWithFallback Should renders correctly with empty src 1`] = `
<DocumentFragment>
  <img
    alt="poster"
    height="16"
    src="test-file-stub"
    width="9"
  />
</DocumentFragment>
`;
